### ISSUE TITLE
What about introducing contextual selectors?

### ISSUE THREAD

**Issue by thierryk** [yahoo/atomizer#44]
We often use rules like this one:
```css
.menu-on .menu-button {
    display: block;
}
```
What if we created atomic classes to take care of this and remove the need of most of these rule?
Imaging this:
```html
<button class="menu-on--D-b" ..>
```
Such class would produce this in atomic.css:
```css
.D-b,
.menu-on .menu-on--D-b {
    display: block;
}
```

   * Pros:
      
      * selectors are grouped, there is no declaration redundancy
      * it's declarative, the logic is in the class itself

   * Cons:
      
      * if the contextual class changes, then authors would need to search and replace it in many places.   

I don't think this would be a good idea after all, but it was worth writing it down :)

**Comment by renatoi** [comment]
As we discussed offline, we will add contextual selectors inspired by [XPath Expressions](http://www.w3.org/TR/xpath-30/#id-path-expressions) in the form of:

**Parent**:

```html
<div class="menu-on"></div>
    <div>
        <button class="menu-on//D-ib"></div>
    </div>
</div>
```

```css
.menu-on .menu-on\/\/D-b {
    display: block;
}
```

**Parent and direct child**:

```html
<div class="menu-on"></div>
    <button class="menu-on/D-ib"></div>
</div>

```css
.menu-on > .menu-on\/D-b {
    display: block;
}
```

**Comment by renatoi** [comment]
We shouldn't encourage this approach though (Maybe we shouldn't even add). But if you are using a component framework, you want your component to not be tightly coupled with stuff outside of its domain. By allowing contextual selectors, we would be allowing this bad practice, while not only making the component less reusable but also the CSS class itself (`.menu-on > .menu-on\/D-b` isn't reusable).

This means that `<Button>` shouldn't know anything about its parent. If it needs to be different based on the context, then the parent component `<Menu>` should control what it does with `<Button>`, by passing additional classes or not.

**Comment by renatoi** [comment]
This would only be useful if you're not using a component-based framework (or no framework at all) and/or don't have much control over the markup.

**Comment by src-code** [comment]
Yeah, as I was saying to @renatoi offline, it feels like this would violate the spirit of Atomic CSS somewhat, in that you could end up bloating the CSS with all these contextual selectors (the combined declarations are like what you'd get out of sass...)

**Comment by src-code** [comment]
> This would only be useful if you're not using a component-based framework (or no framework at all) and/or don't have much control over the markup.

There's something to be said for focusing on a specific problem space we want Atomic to address and own, and I think that's component-based frameworks.  That's Atomic's strength, and we shouldn't water it down to accommodate legacy development practices.

**Comment by thierryk** [comment]
As we discussed offline, we could have this "offering" in a **advance** section. Not something we'd promote but something that would answer some people question: 

> so Atomic cannot do contextual selectors, so dumb!

See what I mean? Besides... who knows? There could be use cases where it could be handy...

**Comment by src-code** [comment]
> As we discussed offline, we could have this "offering" in a advance section. Not something we'd promote but something that would answer some people question:

>> so Atomic cannot do contextual selectors, so dumb!

>See what I mean? Besides... who knows? There could be use cases where it could be handy...

Just because someone thinks its dumb doesn't make them right, does it?

If we don't want people to use it because we're focusing on a problem space that doesn't require them, then I think this is the wrong route to go down, and a waste of valuable time.  Better to open an issue like this, get a conversation going with real users, and then re-evaluate the need for it down the road.

I'm not saying 'never', but for now it seems like an unnecessary diversion.

**Comment by thierryk** [comment]
> Just because someone thinks its dumb doesn't make them right, does it?

No, but providing extra stuff that people *could ask for* somehow shows that we have thought the whole thing through (kinda). Which in my opinion is a good thing as people going through the docs could see this not as something they could use, but something that proves the solution covers a lot of ground.

And once again, this is not something we would promote in any way, could be small print in the advanced section...

**Comment by thierryk** [comment]
ok, one more try... :)

This is a valid use case which would not promote bad practice (to the contrary); when authors want to **reveal** elements on mouseover.

```html
<div class="trigger">
    ...
    <button class="Op-0 Op-1:f trigger:h//Op-1">foo</button>
    <button class="Op-0 Op-1:f trigger:h//Op-1" class="">bar</button>
</div>
```

And the above example shows that - contrary to what I suggested before - there are times where it makes no sense to pair `:hover` with `:focus`, so may be better to make things more **explicit** as it was suggested in another thread yesterday.

I think this is a valid example that makes my case for supporting descendant selectors stronger, no?

**Comment by 3den** [comment]
The context can be easy handled by React, i think this feature would be more usefull if it allowed users some how to write less code:

```
<div class="W-100% col=W-25,Fl-start">
    <div class="col">...</div>
    <div class="col">...</div>
    <div class="col">...</div>
    <div class="col">...</div>
</div>
```

**Comment by thierryk** [comment]
@3den that's *interesting* but not related to what we're discussing here. What you show suggests targeting children instead of the node itself via an ancestor).

We could use a syntax that follows the approach we have taken so far:
```html
<ul class="myList myList<W-25 myList<Fl-start">
    <li>...</li>
    <li>...</li>
    <li>...</li>
    <li>...</li>
</div>
```

```css
.myList > * {
    width: 25%;
    float: $LEFT;
}
```

It is the exact reverse of what we are doing to target ancestors but this serves a *very different purpose* which is DRYNESS in the markup. Something that goes against Atomic philosophy because this would lead to bloating the style sheet because either we will end up with duplicate declarations OR grouped selectors, as it shows below:

```css
.Fl-start {
    float: $LEFT;
}
.W-25\% {
    width: 25%;
}
.myList > * {
    width: 25%;
    float: $LEFT;
}
```

or

```css
.Fl-start,
.myList > * {
    float: $LEFT;
}
.W-25\%,
.myList > * {
    width: 25%;
}
```

**Comment by 3den** [comment]
I like the approach above the d you think it is possible to group the the styles:

```html
<ul class="myList>W-25|Fl-start">
    <li>...</li>
    <li>...</li>
    <li>...</li>
    <li>...</li>
</ul>
```

**Comment by thierryk** [comment]
It's possible but it would create CSS bloat which goes against Atomic's philosophy.
Redundancy in the markup is not as bad as redundancy in the CSS.

**Comment by 3den** [comment]
The css generated could be exactly the same one that you sugested above

**Comment by 3den** [comment]
We just need to make the parser/compiler smarter...

**Comment by thierryk** [comment]
@3den

> The css generated could be exactly the same one that what you suggested above

Yes, and this is the reason we do *not* want to go this route, because we do not want redundancy in the style sheet (declarations or selectors does not matter).

**Comment by thierryk** [comment]
Following our latest offline discussions, the syntax would not follow [XPath Expressions](http://www.w3.org/TR/xpath-30/#id-path-expressions); we would instead use `_` for descendant combinator and `>` for child combinator, like so:

**Descendant selector**:

```html
<div class="menu-on">
    <div>
        <a href="#" class="menu-on_Td-u"></a>
    </div>
</div>
```

```css
.menu-on .menu-on_Td-u {
    text-decoration: underline;
}
```

**Child selector**:

```html
<div class="menu-on">
    <button class="menu-on>D-ib"></div>
</div>
```

```css
.menu-on > .menu-on\>D-ib {
    display: inline-block;
}
```

**Comment by 3den** [comment]
on the example above case above we will have CSS duplication the same way...

```
D-ib {
    display: inline-block;
}

.menu-on > .menu-on\>D-ib {
    display: inline-block;
}

.menu-on .menu-on_D-ib {
     display: inline-block;
}
```

There is duplication on the CSS and HTML... Seems like the decedents and child selectors server the same purpose IMHO the child reads better maybe we can use it for both:

```
D-ib {
    display: inline-block;
}

.menu-on .menu-on\>D-ib {
    display: inline-block;
}
```

But I think it would be nicer to have some way of avoiding code duplication on the markup, that would be very useful to create grids:

```html
<div class="grid>W-1/3|Fl-start|D-ib">
    <div>...</div>
    <div>...</div>
</div>
```

```css
.grid\>W-25%\|Fl-start\|D-ib > * {
    display: inline-block;
    float: $start;
    width: 25%;
}
```

**Comment by 3den** [comment]
IMHO redundancy on the markup is as bad as redundancy in css, because there is no cnd for the markup and it is much harder to cache than CSS.

**Comment by thierryk** [comment]
> on the example above case above we will have CSS duplication the same way...

There is a **big difference** between the 2. We cannot solve the first problem with a plain vanilla approach while we can simply add classes where needed to solve the need of targeting children.

In other words, we "accept" redundancy because there is no other/better way to create classes targeting ancestors. 

Allowing "shortcuts" as you propose could result in outputting rules grouping many different declarations (as your example suggests) which *will* create unnecessary bloat in the style sheet.

**Comment by src-code** [comment]
Addressed in pull #108
